File: hw/paillacho/U3/HW20Composite/CompositePy/model/Supervisor.py
"""
Abstract Supervisor class (can have direct reports)
@author Carlos Paillacho
"""
from abc import ABC
from model.Employee import Employee


class Supervisor(Employee, ABC):
    def __init__(self, name: str = None):
        super().__init__(name)
        self.direct_reports: list[Employee] = []

    def state_name(self):
        # Print this supervisor first
        super().state_name()
        # Then each direct report
        for e in self.direct_reports:
            e.state_name()

    def add(self, an_employee: Employee):
        # Adds a direct report
        self.direct_reports.append(an_employee)

File: hw/paillacho/U3/HW20Composite/Composite/src/main/java/ec/edu/espe/composite/model/Supervisor.java
package ec.edu.espe.composite.model;

import java.util.ArrayList;
import java.util.List;

/**
 *
 * @author Carlos Paillacho
 */
public abstract class Supervisor extends Employee {
    protected List<Employee> directReports = new ArrayList<>();

    /**
     * Prints this supervisor's name and then each direct report's name.
     */
    @Override
    public void stateName() {
        super.stateName();
        if (!directReports.isEmpty()) {
            for (Employee e : directReports) {
                e.stateName();
            }
        }
    }

    /**
     * Adds a direct report to this supervisor.
     *
     * @param anEmployee the employee to add
     */
    public void add(Employee anEmployee) {
        directReports.add(anEmployee);
    }
}
